### PR TITLE
Support LibraryManagedAttributes<TComponent, TAttributes> JSX namespace type

### LINKED ISSUE #23812: Better support for 'defaultProps' in JSX
This proposal aims to fix problems around React's use of `defaultProps`, which TypeScript has trouble understanding today.

# Background

Today, React has a concept called `defaultProps`; these are props that an external user of a component may consider optional, but are *always* present in the component definition.

```ts
export class Greeting extends React.Component {
  static defaultProps = {
    name: 'stranger'
  }

  render() {
    return (
      <div>Hello, {this.props.name}</div>
    )
  }
}

let myElement1 = <Greeting name="Daniel!"/>; // <div>Hello, Daniel!</div>
let myElement2 = <Greeting />;               // <div>Hello, stranger</div>
```

Unfortunately, TypeScript `--strictNullChecks` users who want to take advantage of this are in a tough spot.
They can either specify these props as optional and assert that defaulted properties are present internally,

```ts
export interface GreetingProps {
  name?: string
}

export class Greeting extends React.Component<GreetingProps> {
  static defaultProps = {
    name: 'stranger'
  }

  render() {
    // Notice the non-null assertion!
    //                           | |
    //                           | |
    //                           \ /
    //                            V
    return (
      <div>Hello, {this.props.name!.toUpperCase()}</div>
    )
  }
}
```

or they can jump through levels of indirection:

```ts
export interface GreetingProps {
  name?: string
}

// Notice we are using a 'const' declaration here!
export const Greeting: React.Component<GreetingProps> = class extends React.Component<GreetingProps> {
  static defaultProps = {
    name: 'stranger'
  }

  render() {
    return (
      <div>Hello, {this.props.name!.toUpperCase()}</div>
    )
  }
}
```

# Proposals

## Using resolution semantics of the JSX factory

Users can already emulate this behavior with `React.createElement`: https://tinyurl.com/y9jbptt9

```ts
// React-like API

export interface Component<P> {
    props: P;
}

export interface ComponentClass<P, Defaults = void> {
    new(props: P): Component<P>;
    defaultProps: Defaults;
}

export interface Element<P> {
    type: string | ComponentClass<P>;
    props: P;
}

// Some additions

type Defaultize<T, Defaults> =
    & Partial<Pick<T, Extract<keyof T, keyof Defaults>>>
    & Pick<T, Exclude<keyof T, keyof Defaults>>;

declare function createElement<Props, Defaults>(
    type: ComponentClass<Props, Defaults>,
    props: Defaultize<Props, Defaults> | null):
    Element<Props>

// Now comes your code...

class FooComponent {
    props: { hello: string, world: number };
    static defaultProps: { hello: "hello" };
}

// works: 'hello' got defaulted
createElement(FooComponent, { world: 100 });

// error: missing 'world'
createElement(FooComponent, { hello: "hello" });
```

In theory, just passing the right data to the JSX factory should do the work for us.

However, there are some issues:

1. The `Defaultize` type is hard to read.
2. TypeScript treats the object literal in `createElement` as an inference site, and when inference fails, users get no completions for props. It's not ideal in the general case, but the goal is to continue to give a good experience in JSX.

## Add a new type in the JSX namespace

We could potentially add a new field called `ComponentDefaultPropNames` in `JSX` which would either be a union of literal types or possibly an object type whose property names describe the respective names that `defaultProps` could have:

```ts
namespace JSX {
    export type ComponentDefaultPropNames = "defaultProps";
}
```

When writing out a JSX tag for a component, `JSX.ComponentDefaultPropNames` is looked up on the type of the tag name to determine which properties should be considered optional.

So for example:

```ts
<TagName {...someObject}>
```

If `TagName` expects props with the type `Props`, TypeScript can relate `typeof someObject` to some combination of `Props` and `(typeof TagName)[JSX.ComponentDefaultPropNames]`.

# Potential issues

Under either proposal, *some type* is inferred from the type of the component's `defaultProps` property.
Since [React's `.d.ts` files currently define `defaultProps` to be a `Partial<P>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/01e0ffdbab16b15c702d5b8c87bb122cc6215a59/types/react/index.d.ts#L329), where `P` is the type parameter for the type of props, we need to avoid that from making all props optional.

To avoid issues here, we could say that *only required properties* on `defaultProps` satisfy a JSX element's type. So for example:

```ts
interface MyProps {
  name: string;
  age: number;
}

class MyComponent extends React.Component<MyProps> {
  render() {
    return <div>My name is {this.props.name} and I am {this.props.age} years old.</div>;
  }

  static defaultProps: { name?: string, age: number } = {
    name: "Daniel",
    age: 26,
  }
}

// error: 'name' is still considered required
let x = <MyComponent />

// works! 'age' is optional because it was definitely present in the type of 'defaultProps'.
let y = <MyComponent name="Daniel" />
```

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #23812

Allows, eg:
```ts
class Component extends ReactComponent {
    static propTypes = {
        foo: PropTypes.number,
        bar: PropTypes.node,
        baz: PropTypes.string.isRequired,
    };
    static defaultProps = {
        foo: 42,
    }
}

const a = <Component foo={12} bar="yes" baz="yeah" />;
const b = <Component foo={12} />; // Error, missing required prop bar
const c = <Component bar="yes" baz="yeah" />;
const d = <Component bar="yes" baz="yo" bat="ohno" />; // Error, baz not a valid prop
const e = <Component foo={12} bar={null} baz="cool" />; // bar is nullable/undefinable since it's not marked `isRequired`
const f = <Component foo={12} bar="yeah" baz={null} />; // Error, baz is _not_ nullable/undefinable since it's marked `isRequired`
```

when `JSX.LibraryManagedAttributes<TComponent, TAttributes>` is appropriately defined. Thanks to conditional types, this method of supporting `propTypes` and `defaultProps` allows definition authors a ton of freedom in how the static types need to be transformed before they are checked against what the user wrote. (Allowing customization of, for example: how conflicts between provided props and inferred props are handled, how inferences are mapped, how optionality is handled, and how inferences from differing places should be combined)

As an aside, you can fully implement the functionality of both `InstrinsicAttributes` and `InstrinsicClassAttributes` using this new type entrypoint, so in a way it obsoletes them.

Feel free to 🚲 🏠  with the name and give feedback; the actual implementation is relatively simple. @DanielRosenwasser you probably have some opinions here, right? 😄

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
Could you elaborate in this PR on what `LibraryManagedAttributes` is intended to do and give some specific implementations examples?

File: tests/baselines/reference/tsxLibraryManagedAttributes.errors.txt [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  Type 'number' is not assignable to type 'string | undefined'.
+
+
+==== tests/cases/conformance/jsx/tsxLibraryManagedAttributes.tsx (15 errors) ====
+    type Defaultize<TProps, TDefaults> =
+        & {[K in Extract<keyof TProps, keyof TDefaults>]?: TProps[K]}
+        & {[K in Exclude<keyof TProps, keyof TDefaults>]: TProps[K]}
+        & Partial<TDefaults>;
+    
+    type InferedPropTypes<P> = {[K in keyof P]: P[K] extends PropTypeChecker<infer T, infer U> ? PropTypeChecker<T, U>[typeof checkedType] : {}};
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/tsxLibraryManagedAttributes.errors.txt]: I know it's just a test, but `InferedPropTypes` -> `InferredPropTypes`

File: tests/baselines/reference/tsxLibraryManagedAttributes.js [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+class JustDefaultPropsWithSpecifiedGeneric extends ReactComponent<FooProps> {
+    static defaultProps = {
+        foo: "no",
+    };
+}
+
+const x = <JustDefaultPropsWithSpecifiedGeneric foo="eh" />;
+const y = <JustDefaultPropsWithSpecifiedGeneric foo="no" bar="ok" />; // error, no prop named bar
+const z = <JustDefaultPropsWithSpecifiedGeneric foo={12} />; // error, wrong type
+
```
Comment (jaredpalmer) [inline tests/baselines/reference/tsxLibraryManagedAttributes.js]: What about `<JustDefaultPropsWithSpecifiedGeneric  />`? Is that kosher?

**Comment by weswigham** [comment]
@jaredpalmer The `Props` type parameter should be the "internal" props (so what you read), while the props returned by the type alias should be the "external" props (the ones you write in a tag). So, instead of
```ts
interface Props {
  thing?: string; // still optional
}

class Thing extends React.Component<Props> {
   static defaultProps = {
      thing: 'hello'
   }

  render() {
    console.log(this.props.thing) // inferred because of defaultProps, no ! needed
    // ...
 }
}
// ..
<Thing /> // works
```
you'd have
```ts
interface Props {
  thing: string; // note: not optional within class, but is optional when writing due to the default
}

class Thing extends React.Component<Props> {
   static defaultProps = {
      thing: 'hello'
   }

  render() {
    console.log(this.props.thing) // inferred because of defaultProps, no ! needed
    // ...
 }
}
// ..
<Thing /> // works
``` 
or at least that's what makes sense to me - since the `Props` generic param is the type of the first argument to the factory/class (which includes all defaults).

**Comment by jaredpalmer** [comment]
got it so it will work like Flow. cool!

**Comment by Hotell** [comment]
This is indeed awesome! In the meantime I wrote an article for React consumers struggling with this "deal breaker" ✌️ https://medium.com/@martin_hotell/react-typescript-and-defaultprops-dilemma-ca7f81c661c7

File: tests/cases/conformance/jsx/tsxLibraryManagedAttributes.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
+// @jsx: preserve
+// @strict: true
+
+type Defaultize<TProps, TDefaults> =
+    & {[K in Extract<keyof TProps, keyof TDefaults>]?: TProps[K]}
+    & {[K in Exclude<keyof TProps, keyof TDefaults>]: TProps[K]}
+    & Partial<TDefaults>;
```
Comment (ferdaber) [inline tests/cases/conformance/jsx/tsxLibraryManagedAttributes.tsx]: Why is the `Partial<TDefaults>` here intersected with the extracted mapped type from `TDefaults`? What kind of error does that produce if `typeof TProps[K]` is incompatible with `typeof TDefaults[K]`?

**Comment by weswigham** [comment]
@mhegazy I've synchronized this branch with `master`; can I get this reviewed?

**Comment by mhegazy** [comment]
We will need to update https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/JSX.md

**Comment by OliverJAsh** [comment]
What is the expected behaviour when the props type is a union? Is this expected?
``` tsx
{
    type Props = { shared: string } & (
        | {
              type: 'foo';
              foo: string;
          }
        | {
              type: 'bar';
              bar: string;
          });
    class Component extends ReactComponent<Props> {
        static defaultProps = {
            shared: 'yay',
        };
    }

    <Component type="foo" />; // Expected error, got none
    // Unexpected error
    // Property 'foo' does not exist on type 'Defaultize<Props, { shared: string; }>'.
    <Component type="foo" foo="1" />;
}
```

**Comment by OliverJAsh** [comment]
@mhegazy Is it an issue worth tracking? If so, I'll open a separate issue.

**Comment by mhegazy** [comment]
yes please.

**Comment by weswigham** [comment]
@OliverJAsh it's probably also worth noting that while this code is in `typescript`, there won't be any changes until the `react` `.d.ts` is updated to use the new jsx namespace entrypoint; so until that happens `defaultProps` still won't do anything.

**Comment by OliverJAsh** [comment]
@mhegazy https://github.com/Microsoft/TypeScript/issues/25503.

**Comment by DanielRosenwasser** [comment]
For any questions/issues, please file a separate follow-up issue.

**Pull request merged**